src/lib/matchers.rs

```diff
@@ -216,19 +281,264 @@ impl<T: Clone> MappingStore<T> {
     /// To nodes are common if they have already been matched.
     fn num_common_descendants(&self, from: &NodeId, to: &NodeId) -> u32 {
         let mut dst_desc = HashSet::new();
-        for node in to.breadth_first_traversal(&self.to) {
+        for node in to.breadth_first_traversal(&self.to_arena) {
             dst_desc.insert(node);
         }
         let mut common = 0;
         let mut to: Option<NodeId>;
-        for node in from.descendants(&self.from) {
+        for node in from.descendants(&self.from_arena) {
             to = self.get_to(&node);
             if to.is_some() && dst_desc.contains(&to.unwrap()) {
                 common += 1;
             }
         }
         common
     }
+
+    /// Given a mapping store, generate an edit script.
+    ///
+    /// This function implements the optimal algorithm of Chawathe et al. (1996).
+    /// Variable names as in Figures 8 and 9 of the paper.
+    pub fn generate_edit_script(&mut self, root: NodeId) -> EditScriptResult<T> {
+        let mut script: EditScript<T> = EditScript::new();
+        let mut from_in_order: HashSet<NodeId> = HashSet::new();
+        let mut to_in_order: HashSet<NodeId> = HashSet::new();
+        let mut next_from_id = self.from_arena.size();
+        let mut new_mappings = TemporaryMappingStore::new();
+        // Copy current mappings over to new_mappings.
+        for (key, value) in &self.from {
+            new_mappings.push(*key, value.0);
+        }
+        // Combined update, insert, align and move phases.
+        let tmp_to_arena = self.to_arena.clone();
+        for x in root.breadth_first_traversal(&tmp_to_arena) {
+            if x.is_root(&self.to_arena) {
```

Comment:
I'm not sure, if I say:

```rust
for x in root.breadth_first_traversal(&tmp_to_arena).filter(|n| !n.is_root(&self.to_arena))
```

I get:

```rust
error[E0502]: cannot borrow `self.from_arena` as mutable because `self` is also borrowed as immutable
   --> src/lib/matchers.rs:333:32
    |
313 |         for x in root.breadth_first_traversal(&tmp_to_arena).filter(|n| !n.is_root(&self.to_arena)) {
    |                                                                     ---             ---- previous borrow occurs due to use of `self` in closure
    |                                                                     |
    |                                                                     immutable borrow occurs here
...
333 |                 ins.apply(&mut self.from_arena)?;
    |                                ^^^^^^^^^^^^^^^ mutable borrow occurs here
...
377 |         }
    |         - immutable borrow ends here
```